scriptworker/artifacts.py

```diff
@@ -0,0 +1,303 @@
+
+
+_GZIP_SUPPORTED_CONTENT_TYPE = ('text/plain', 'application/json', 'text/html', 'application/xml')
+_EXTENSIONS_TO_FORCE_TO_PLAIN_TEXT = ('.asc', '.log')
+
+
+def _force_mimetypes_to_plain_text():
+    """Populate/Update the mime types database with supported extensions that we want to map to 'text/plain'.
+
+    These extensions can then be open natively read by browsers once they're uploaded on S3. It doesn't affect artifacts
+    once they're downloaded from S3.
+    """
+    for extension in _EXTENSIONS_TO_FORCE_TO_PLAIN_TEXT:
+        mimetypes.add_type('text/plain', extension)
+        log.debug('Extension "{}" forced to text/plain'.format(extension))
+
+
+_force_mimetypes_to_plain_text()
+
+
+# upload_artifacts {{{1
+async def upload_artifacts(context):
+    """Compress and upload the files in ``artifact_dir``, preserving relative paths.
+
+    Compression only occurs with files known to be supported.
+
+    This function expects the directory structure in ``artifact_dir`` to remain
+    the same.  So if we want the files in ``public/...``, create an
+    ``artifact_dir/public`` and put the files in there.
+
+    Args:
+        context (scriptworker.context.Context): the scriptworker context.
+
+    Raises:
+        Exception: any exceptions the tasks raise.
+    """
+    file_list = {}
+    for target_path in filepaths_in_dir(context.config['artifact_dir']):
+        path = os.path.join(context.config['artifact_dir'], target_path)
+
+        content_type, content_encoding = compress_artifact_if_supported(path)
+        file_list[target_path] = {
+            'path': path,
+            'target_path': target_path,
+            'content_type': content_type,
+            'content_encoding': content_encoding,
+        }
+
+    tasks = []
+    for upload_config in file_list.values():
+        tasks.append(
+            asyncio.ensure_future(
+                retry_create_artifact(
+                    context, upload_config['path'],
+                    target_path=upload_config['target_path'],
+                    content_type=upload_config['content_type'],
+                    content_encoding=upload_config['content_encoding'],
+                )
+            )
+        )
+    await raise_future_exceptions(tasks)
+
+
+def compress_artifact_if_supported(artifact_path):
+    """Compress artifacts with GZip if they're known to be supported.
+
+    This replaces the artifact given by a gzip binary.
+
+    Args:
+        artifact_path (str): the path to compress
+
+    Returns:
+        content_type, content_encoding (tuple):  Type and encoding of the file. Encoding equals 'gzip' if compressed.
+    """
+    content_type, encoding = guess_content_type_and_encoding(artifact_path)
+    log.debug('"{}" is encoded with "{}" and has mime/type "{}"'.format(artifact_path, encoding, content_type))
+
+    if encoding is None and content_type in _GZIP_SUPPORTED_CONTENT_TYPE:
+        log.debug('"{}" can be gzip\'d. Compressing...'.format(artifact_path))
```

Comment:
I'm good with turning this line into info. I'd keep the final log info as well. Otherwise let's say an error occurs after gziping, then I'm afraid people may wonder if files were gzip'd in the end of not. This may lead them to wrong assumptions about what's failing.